kernel/src/sched.rs

```diff
@@ -524,6 +520,55 @@ impl Kernel {
         }
     }
 
+    /// Main loop of the OS.
+    ///
+    /// Most of the behavior of this loop is controlled by the `Scheduler`
+    /// implementation in use.
+    pub fn kernel_loop<P: Platform, C: Chip, SC: Scheduler<C>, const NUM_PROCS: usize>(
+        &self,
+        platform: &P,
+        chip: &C,
+        ipc: Option<&ipc::IPC<NUM_PROCS>>,
+        scheduler: &SC,
+        capability: &dyn capabilities::MainLoopCapability,
+    ) -> ! {
+        chip.watchdog().setup();
+        loop {
+            self.loop_op(platform, chip, ipc, scheduler, capability, false);
+        }
+    }
+
+    /// Main test loop of the OS.
+    ///
+    /// This is similar to `kernel_loop()` but also calls tests passed in via
+    /// the `tests` argumet. This should only be used for running tests.
+    pub fn test_kernel_loop<P: Platform, C: Chip, SC: Scheduler<C>, const NUM_PROCS: usize>(
```

Comment:
Not that I'm one to advocate for more `cfg`s, but at the same time, this function should _only_ ever be used in a CI/test setup, right? Feels like we should indicate that.

This does also break our prior policy of only `cfg`ing arch-specific stuff. It sounds like the tradeoffs here have been discussed a fair bit by the OT-WG, and `cfg`ing test harnesses is pretty well-supported by Rust and is a generally supported notion in the ecosystem, but would like to understand whether this really needs to be a separate function, or if all we _really_ need is to expose a sleep / don't sleep parameter?

Reply:
I originally had this function protected by a `#[cfg(not(test))]`, but Hudson preferred that to be removed.

I know Tock doesn't like `cfg`, but `cfg(test)` is pretty core to Rust. It's support by all of the Rust tools and used by most Rust projects so in this case I don't think we are doing anything too confusing here.

Like I mentioned in my other reply, we want to run the tests as part of the loop AND we also want the kernel loop to end, so we can then report success. That's why there is a change to the kernel loop.

Reply:
I think people hacking around might want to be able to run the test version without having to go through layers of rust testing cfgs and infrastructure.